pkg/service/rpc/race.go

```diff
@@ -0,0 +1,52 @@
+package rpc
+
+import (
+	"context"
+	"sync/atomic"
+	"unsafe"
+)
+
+type raceResult[T any] struct {
+	i   int
+	val *T
+	err error
+}
+
+type Race[T any] struct {
+	ctx       context.Context
+	cancel    context.CancelFunc
+	nextIndex int
+	res       unsafe.Pointer
+}
+
+// NewRace creates a race to yield the result from one or more candidate
+// functions
+func NewRace[T any](ctx context.Context) *Race[T] {
+	ctx, cancel := context.WithCancel(ctx)
+	return &Race[T]{
+		ctx:    ctx,
+		cancel: cancel,
+	}
+}
+
+// Go adds a candidate function to the race by running it in a new goroutine
+func (r *Race[T]) Go(fn func(ctx context.Context) (*T, error)) {
+	i := r.nextIndex
+	r.nextIndex++
+
+	go func() {
+		val, err := fn(r.ctx)
+		atomic.CompareAndSwapPointer(&r.res, nil, (unsafe.Pointer)(&raceResult[T]{i, val, err}))
```

Comment:
Is there any strong motivation to use atomic semantics here? A simple mutex may be more approachable and maintainable, particularly as it would avoid using unsafe.Pointer whose semantics can be tricky to get right. 

To quote the documentation of the go atomic package (https://pkg.go.dev/sync/atomic):

> These functions require great care to be used correctly. Except for special, low- level applications, synchronization is better done with channels or the facilities of  the sync package. Share memory by communicating; don't communicate by sharing memory.

Reply:
we can replace this with atomic.Pointer if we can use go 1.19. the implementation here is correct and i think it's well enough contained in the helper (as opposed to out in service code or wherever). i think this is a temporary workaround that we'll want to remove after the migration. also go/the docs are pretty patronizing so i would take that warning with a grain of salt. i can add a unit test if it would make you more comfortable.

Reply:
What is the pro of using CompareAndSwap or a construct that requires go 1.19 (which I think we do not want to do unless we have very good reason)? Most folks reading the code will need to refer to the documentation to understand the exact semantics. I had to do so to make sure that your usage was correct.

I am in full agreement with the go documentation here: unless in a performance critical path, code is best optimized to be easy to read, and document clearly intent. This is especially true for open source code where outside contributions are encouraged.

Reply:
keep in mind this is temporary for the switch to psrpc and will be removed in a few versions. I think we definitely wouldn't want to upgrade to 1.19 for it

Reply:
the pro is that it's the simplest thing that does what the function is trying to accomplish. go is anemic enough without restricting use of an entire class of synchronization primitives. and this wouldn't make a list of the top 10k things that make working working with this codebase challenging.